dlrover/python/master/diagnosis/precheck_operator.py

```diff
@@ -44,39 +59,269 @@ def get_retry_interval_secs(cls) -> int:
+        start = time.time()
+        while True:
+            if time.time() - start > timeout:
+                logger.warning(
+                    f"Scheduling hasn't started for over {timeout}s."
+                )
+                return False
+
+            has_started = False
+            job_nodes = job_ctx.job_nodes()
+            for _, nodes in job_nodes.items():
+                for _, node in nodes.items():
+                    if node.create_time:
+                        has_started = True
+            if has_started:
+                return True
+            else:
+                logger.info(
+                    f"Scheduling hasn't started yet, wait {wait_time}s..."
+                )
+                time.sleep(wait_time)
+
+    def check(self, *args, **kwargs):
+        job_args: JobArgs = kwargs.get("job_args")
+        job_type = job_args.distribution_strategy
+        strategy = _dlrover_ctx.pending_fail_strategy
+        timeout = get_pending_timeout()
+        logger.info(
+            f"{self.__class__.__name__} do checking with "
+            f"job_type: {job_type}, "
+            f"strategy: {strategy}, "
+            f"timeout: {timeout}"
+        )
+
+        if timeout <= 0 or strategy == PendingTimeoutStrategyType.SKIP:
+            msg = (
+                f"Skip {self.__class__.__name__} for "
+                "'skip' pending timeout strategy."
+            )
+            logger.info(msg)
+            return PreCheckResult(result_msg=msg)
+
+        if not self.wait_scheduling_started():
+            logger.warning("All nodes pending when scheduling.")
+            return PreCheckResult(
+                result=1,
+                result_msg=SchedulingPreCheckOperator.SCHEDULING_FAILED_MSG,
+                abnormal_nodes=[],
+            )
+
+        round = 0
+        while True:
+            logger.info(f"Scheduling pre-check round: {round}")
+            if job_type == DistributionStrategy.ALLREDUCE:
+                cur_nodes = list(
+                    job_ctx.job_nodes_by_type(NodeType.WORKER).values()
+                )
+                pending_result = self.check_allreduce_job_pending(
+                    cur_nodes, timeout, strategy
+                )
+            elif job_type == DistributionStrategy.PS:
+                ps_nodes = list(
+                    job_ctx.job_nodes_by_type(NodeType.PS).values()
+                )
+                worker_nodes = list(
+                    job_ctx.job_nodes_by_type(NodeType.WORKER).values()
+                )
+                pending_result = self.check_ps_job_pending(
+                    ps_nodes + worker_nodes, timeout, strategy
+                )
+            else:
+                msg = f"Skip {self.__class__.__name__} for {job_type}."
+                logger.warning(msg)
+                return PreCheckResult(result_msg=msg)
+
+            if pending_result[0]:
+                # has pending
```

Comment:
`has pending` -> `job is pending`

Reply:
Is node pending, we don't have the concept of 'job pending'.